File: packages/core/src/Tgrok.ts
import { EventEmitter } from "@drmer/core";
import { ControlClient } from "./ControlClient";
import { Log } from "./Log";
import { Tunnel } from "./Tunnel";
import { event } from "./event";

interface TgrokStatus {
  status: number,
  tunnels: any[],
}

class Tgrok extends EventEmitter {
  public host: string = "ngrok.local";
  public port: number = 4443;
  public config: any = {};
  public context: any = {};

  private started = false;

  private controlClient?: ControlClient;
  private timerId: any;
  private retryTimes = 0;

  public set debug(v: boolean) {
    Log.debug = v;
  }

  /**
   * Deprecated, use `start` in favor.
   */
  public startLocal = (lport?: number | string, domain?: string): void => {
    this.start(lport, domain);
  };

  public start = (lport?: number | string | Tunnel[], domain?: string): void => {
    // no repeated start
    if (this.started) {
      Log.error("\n\tAlready Started!!!\n");

      return;
    }
    this.started = true;

    let tunnels: Tunnel[] = [];

    if (typeof lport === "object") {
      tunnels = lport as Tunnel[];
    } else {
      let localPort = 80;
      let subdomain = domain;

      if (typeof lport === "number") {
        localPort = lport;
      } else {
        subdomain = lport as string;
      }
      tunnels.push(new Tunnel({
        protocol: "http",
        hostname: "",
        subdomain,
        rport: 0,
        lhost: "127.0.0.1",
        lport: localPort,
      }));
    }

    const client = new ControlClient(this.config, tunnels);

    client.on("connect", this.onConnect);
    client.on("error", this.onError);
    client.on("end", this.onEnd);
    this.controlClient = client;
    this.connect();
  };

  protected connect = (): void => {
    if (!this.controlClient) {
      return;
    }
    this.controlClient.start(this.host, this.port, this.context);
  };

  protected onConnect = (): void => {
    this.retryTimes = 0;
  };

  protected onEnd = (): void => {
    event.emit("info", {
      evt: "master:error",
      payload: `reconnect after ${this.timeout}s`,
    });
    Log.error(`main socket onEnd, reconnect after ${this.timeout}s`);
    this.reconnect(false);
  };

  protected onError = (_err: Error): void => {
    event.emit("info", {
      evt: "master:error",
      payload: `reconnect after ${this.timeout}s`,
    });
    Log.error(`main socket onError, reconnect after ${this.timeout}s`);
    this.reconnect(false);
  };

  public reconnect = (clear: boolean): void => {
    if (clear) {
      clearTimeout(this.timerId);
      this.retryTimes = 0;
      this.timerId = undefined;
    }
    // master socket run into a problem.
    if (this.timerId) {
      // has already restart
      return;
    }
    this.timerId = setTimeout(() => {
      this.timerId = undefined;
      this.connect();
    }, this.timeout * 1000);
    this.retryTimes += 1;
  };

  private get timeout() {
    const timeList = [1, 1, 2, 3, 5, 8, 13, 21];

    if (this.retryTimes >= timeList.length) {
      return timeList[timeList.length - 1];
    }

    return timeList[this.retryTimes];
  }

  public openTunnel = (tunnel: Tunnel): void => {
    if (!this.controlClient) {
      return;
    }
    this.controlClient.openTunnel(tunnel);
  };

  public closeTunnel = (id: string): void => {
    if (!this.controlClient) {
      return;
    }
    this.controlClient.closeTunnel(id);
  };

  public removeTunnel = (id: string): void => {
    if (!this.controlClient) {
      return;
    }
    this.controlClient.removeTunnel(id);
  };

  public get status(): TgrokStatus {
    if (!this.controlClient) {
      return {
        status: 0,
        tunnels: [],
      };
    }
    const tunnels: any[] = [];

    this.controlClient.tunnelList.forEach((tunnel) => {
      tunnels.push({
        id: tunnel.id,
        status: tunnel.status,
      });
    });

    return {
      status: this.controlClient.status,
      tunnels,
    };
  }
}

export {
  Tgrok,
};

File: packages/ios/Tgrok/tgrok/Tgrok.swift
//
//  Tgrok.swift
//  Tgrok
//
//  Created by Yueyu Zhao on 2020/9/13.
//  Copyright © 2020 Yueyu Zhao. All rights reserved.
//

import Foundation
import SwiftyJSON

class Tgrok: NSObject {
    
    var controlClient: ControlClient?
    
    private var timer: Timer?
    private var retryTimes = 0
    
    override init() {
        super.init()
    }
    
    func boot(_ host: String, _ port: UInt16) {
        let controlClient = ControlClient(host, port)
        controlClient.delegate = self
        
        self.controlClient = controlClient
    }
    
    func start(_ port: UInt16, _ host: String?) {
        controlClient?.tunnels.append(Tunnel(JSON([
            "id": UUID().uuidString,
            "protocol": "http",
            "hostname": "",
            "subdomain": "test",
            "remotePort": 0,
            "localHost": host ?? "127.0.0.1",
            "localPort": port,
        ])))
        self.connect()
    }
    
    func start(_ tunnels: [Tunnel]) {
        controlClient?.tunnels = tunnels
        self.connect()
    }

    func openTunnel(_ tunnel: Tunnel) {
        controlClient?.openTunnel(tunnel)
    }
    
    func closeTunnel(_ id: String) -> Bool {
        _ = controlClient?.closeTunnel(id)
        return true
    }
    
    func removeTunnel(_ id: String) -> Bool {
        _ = controlClient?.removeTunnel(id)
        return true
    }
    
    func status() -> JSON {
        var tunnels: [JSON] = []
        if let client = controlClient {
            for tunnel in client.tunnels {
                tunnels.append(JSON([
                    "id": tunnel.id,
                    "status": tunnel.status,
                ]))
            }
        }
        
        return JSON([
            "status": controlClient?.status ?? 0,
            "tunnels": tunnels,
        ])
    }
    
    private func connect() {
        controlClient?.start()
    }
    
    func reconnect(_ clear: Bool) {
        if clear {
            timer?.invalidate()
            self.retryTimes = 0
            self.timer = nil
        }
        
        if self.timer != nil {
            return
        }
        
        let timeout = self.timeout()
        timer = Timer.scheduledTimer(withTimeInterval: timeout, repeats: false, block: { (_) in
            self.timer?.invalidate()
            self.timer = nil
            self.connect()
        })
        self.retryTimes += 1
    }
    
    func timeout() -> Double {
        let times = [1, 1, 2, 3, 5, 8, 13, 21]
        if self.retryTimes >= times.count {
            return Double(times.last!)
        }
        return Double(times[self.retryTimes])
    }
    
}

extension Tgrok : ControlClientDelegate {
    
    func onConnect() {
        self.retryTimes = 0
    }
    
    func onError() {
        let timeout = self.timeout()
        NotificationCenter.default.post(name: .tgrok, object: JSON([
            "evt": "master:error",
            "payload": "reconnect after \(timeout)s"
        ]))
        print("main socket error, reconnect after \(timeout)s")
        self.timer = nil
        self.reconnect(false)
    }
    
}
